pkg/vtpm/src/server.cpp

```diff
@@ -33,12 +41,160 @@ google::protobuf::uint32 readHdr(char *buf)
+    coded_output->WriteVarint32(response.ByteSize());
+    response.SerializeToCodedStream(coded_output);
+    int sent_bytes = send(sock, (void *)resp_buffer, size, 0);
+    delete(resp_buffer);
+    if (sent_bytes == size) {
+      return 0;
+    } else {
+      return -1;
+    }
+}
+
+static inline bool
+isCommandAllowed(string command_with_args)
+{
+  istringstream ss(command_with_args);
+  string command;
+  ss >> command;
+  auto it = find(allowed_commands.begin(), allowed_commands.end(), command);
+  if (it != allowed_commands.end()) {
+     return true;
+  }
+  return false;
+}
+
+static inline bool
+isFileNameAPath(string filename)
+{
+    return (filename.find('/') != filename.npos);
+}
+
+static int
+sanitizeCmdRequest(int sock, eve_tools::EveTPMRequest &request,
+                   eve_tools::EveTPMResponse &response)
+{
+    if (!isCommandAllowed(request.command())) {
+        cout << "Not a legal command, bailing out" << std::endl;
+        response.set_response("Command forbidden!");
+        sendResponse(sock, response);
+        return -1;
+    }
+    for (int i=0; i < request.expectedfiles_size(); i++) {
+        std::string expectedFile = request.expectedfiles(i);
+        if (isFileNameAPath(expectedFile)) {
+        response.set_response("output filename should not be a path!");
+        sendResponse(sock, response);
+        return -1;
+        }
+    }
+    for (int i=0; i < request.inputfiles_size(); i++) {
+        const eve_tools::File& file = request.inputfiles(i);
+        if (isFileNameAPath(file.name())) {
+            response.set_response("input filename should not be a path!");
+            sendResponse(sock, response);
+            return -1;
+        }
+    }
+    return 0;
+}
+
+//TBD: Change client library to send args in a list, rather in a single string
+//till then, split the given string into command and args. Once split, launch
+//the command using execve() inside a child. Parent waits for the client(i.e. command)
+//to finish.
+static int
+execCmd (string cmd) {
+    int i = 0;
+    string command_alone;
+    istringstream full_cmd(cmd);
+    const char *path[MAX_ARGS];
+    string args[MAX_ARGS];
+
+    full_cmd >> command_alone;
+    string command_with_path = BIN_PATH + command_alone;
+
+    path[i++] = command_with_path.c_str();
+    while (full_cmd >> args[i] && i < MAX_ARGS) {
+        path[i] = args[i].c_str();
+        i++;
+    }
+    path[i] = NULL;
+
+    //spawn child process
+    int child_pid = fork();
+    if (child_pid < 0) {
+        perror("fork failed with:");
+        return -1;
+    } else if (child_pid == 0) {
+        //Redirect stdout and stderr to cmd.output file
+        int fd = open(CMD_OUTPUT_FILE, O_CREAT|O_TRUNC|O_WRONLY, 0600);
+        dup2(fd, 1);
+        dup2(fd, 2);
+        close(fd);
+
+        //Flush pending stderr and stdout queues
+        fflush(stderr);
+        fflush(stdout);
+        execve(path[0], (char **)&path, NULL);
```

Comment:
Why not check for it returning error aka < 0 and then reporting errno to caller?

Reply:
fixed it in the latest diff.